archive/100-theorems-list/57_herons_formula.lean

```diff
@@ -0,0 +1,118 @@
+/-
+Copyright (c) 2020 Matt Kempster. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Matt Kempster
+-/
+import geometry.euclidean.triangle
+
+/-!
+# Freek № 57: Heron's Formula
+
+This file proves Theorem 57 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/),
+also known as Heron's formula, which gives the area of a triangle based only on its three sides'
+lengths.
+
+## References
+
+* https://en.wikipedia.org/wiki/Herons_formula
+
+-/
+
+open euclidean_geometry
+open_locale real euclidean_geometry
+
+notation `√` := real.sqrt
+
+variables {V : Type*} {P : Type*} [inner_product_space ℝ V] [metric_space P]
+    [normed_add_torsor V P]
+include V
+
+/-- The sine of a real number `x` is the square root of one minus the cosine squared of `x`. -/
+lemma sin_eq_sqrt_one_minus_cos_sq (x : ℝ) (hl : x ≥ 0) (hu : x ≤ π) :
+  real.sin x = √ (1 - (real.cos x)^2) :=
+begin
+  rw ← real.sqrt_sqr (real.sin_nonneg_of_nonneg_of_le_pi hl hu),
+  exact congr_arg real.sqrt (real.sin_square x),
+end
```

Comment:
This deserves to be in mathlib proper. Maybe in `analysis.special_functions.trigonometric` (along with the `cos` variant)?

Reply:
@bryangingechen, two questions:
1) what do you think the exact formulation of this lemma should be? That is, this is a generalization of the fact that `|sin x| = sqrt(1 - cos^2 x)` alongside `|sin x| = sin x` whenever x is in `[0, pi]`. Should we keep the `[0, pi]` as part of the lemma or not?
2) do you think `data/complex/exponential` is a good place for this lemma, alongside `sin_sq_add_cos_sq`? Or is `trigonometric.lean` better because it's more "specialized"?
cc  @benjamindavidson

Reply:
These are good questions, and I haven't looked at the trigonometric files recently. I think I'd defer to @benjamindavidson's opinion here.

Reply:
I would do the following:
```
lemma abs_sin_eq_sqrt_one_minus_cos_sq (x : ℝ) :
  abs (real.sin x) = √ (1 - real.cos x ^ 2) :=
by rw [← real.sin_square, real.sqrt_sqr_eq_abs]

lemma sin_eq_sqrt_one_minus_cos_sq {x : ℝ} (hl : 0 ≤ x) (hu : x ≤ π) :
  real.sin x = √ (1 - real.cos x ^ 2) :=
by rw [← abs_sin_eq_sqrt_one_minus_cos_sq, abs_of_nonneg (real.sin_nonneg_of_nonneg_of_le_pi hl hu)]

lemma abs_cos_eq_sqrt_one_minus_sin_sq (x : ℝ) :
  abs (real.cos x) = √ (1 - real.sin x ^ 2) :=
by rw [← real.cos_square', real.sqrt_sqr_eq_abs]

lemma cos_eq_sqrt_one_minus_sin_sq {x : ℝ} (hl : -(π / 2) ≤ x) (hu : x ≤ π / 2) :
  real.cos x = √ (1 - real.sin x ^ 2) :=
by rw [← abs_cos_eq_sqrt_one_minus_sin_sq, abs_of_nonneg (cos_nonneg_of_nonneg_of_le_of_le hl hu)]
```
Note that this will require you to also add the lemma (though perhaps with a better name)
```
lemma cos_nonneg_of_nonneg_of_le_of_le {x : ℝ} (hl : -(π / 2) ≤ x) (hu : x ≤ π / 2) :
  0 ≤ real.cos x :=
real.cos_nonneg_of_mem_Icc ⟨hl, hu⟩
```
which probably should have existed already, anyway.


Regarding where they belong:
Since they are just algebraic rearrangements of `sin_square`/`cos_square'`, I would like to see the `abs` lemmas go in `/exponential` right after those lemmas. The `nonneg` versions, though, rely on lemmas that live in `/trigonometric`, so they will have to go in that file.